keps/sig-api-machinery/3157-watch-list/README.md

```diff
@@ -0,0 +1,991 @@
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Today informers are the primary source of LIST requests.
+The LIST is used to get a consistent snapshot of data currently in etcd to build up a client-side in-memory cache.
+The primary issue with LIST requests is unpredictable memory consumption.
+The actual usage depends on many factors like the page size, applied filters (e.g. label selectors), query parameters, and sizes of individual objects.
+See the [Appendix](#appendix) section for more details on potential sources of LIST request and their impact on memory.
+![memory consumption during the synthetic test](./memory_usage_before.png "memory consumption during the synthetic test")
+In extreme cases, the server can allocate hundreds of megabytes per request.
+To better visualize the issue let's consider the above graph.
+It shows the memory usage of an API server during a test (see [manual test](#manual-testing-without-the-changes-in-place) section for more details).
+We can see that increasing the number of informers drastically increases the memory consumption of the server.
+Moreover, around 16:40 we lost the server after running 16 informers. During an investigation, we realized that the server allocates a lot of memory for handling LIST requests.
+In short, it needs to bring data from the database, unmarshal it, do some conversions and prepare the final response for the client.
+The bottom line is around O(5*the_response_from_etcd) of temporary memory consumption.
+Neither priority and fairness nor Golang garbage collection is able to protect the system from exhausting memory.
+
+A situation like that is dangerous twofold.
+First, as we saw it could slow down if not fully stop an API server that has received the requests.
+Secondly, a sudden and uncontrolled spike in memory consumption will likely put pressure on the node itself.
+This might lead to thrashing, starving, and finally losing other processes running on the same node, including kubelet.
+Stopping kubelet has serious issues as it leads to workload disruption and a much bigger blast radius.
+Note that in that scenario even clusters in an HA setup are affected.
+
+Worse, in rare cases (see the [Appendix](#appendix) section for more) recovery of large clusters with therefore many kubelets and hence informers for pods, secrets, configmap can lead to a very expensive storm of LISTs.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+- protect kube-apiserver and its node against list-based OOM attacks
```

Comment:
it doesn't sound like you can protect against attacks (someone intentionally listing).  Your proposal appears to be about allowing well-behaved clients (informers) to change their client code to issue a watch instead.

Reply:
To a degree also against attacks. If you know your system does not need many lists, you can heavily throttle (p&f). Today we cannot because everybody has to list.